test_compileall2.py

```diff
@@ -384,6 +384,84 @@ def test_ignore_symlink_destination(self):
         self.assertTrue(os.path.isfile(allowed_bc))
         self.assertFalse(os.path.isfile(prohibited_bc))
 
+    def test_hardlink_deduplication_bad_args(self):
+        with self.assertRaises(ValueError):
+            compileall.compile_dir(self.directory, quiet=True, optimize=0, hardlink_dupes=True)
+
+    def test_hardlink_deduplication_same_bytecode_all_opt(self):
+        """Same bytecode produced for all levels of optimization"""
+        path = os.path.join(self.directory, "test", "same_all")
+        os.makedirs(path)
+
+        simple_script = script_helper.make_script(path, "test_same_bytecode", "a = 0")
+        pyc_opt0 = importlib.util.cache_from_source(simple_script)
+        pyc_opt1 = importlib.util.cache_from_source(simple_script, optimization=1)
+        pyc_opt2 = importlib.util.cache_from_source(simple_script, optimization=2)
+
+        compileall.compile_dir(path, quiet=True, optimize=[0, 1, 2], hardlink_dupes=True)
+        self.assertTrue(os.stat(pyc_opt0).st_ino == os.stat(pyc_opt1).st_ino == os.stat(pyc_opt2).st_ino)
```

Comment:
I suggest having individual assertEquals here for better failure reading.

Reply:
Indeed. It'll also open some space for more comments.